### PR TITLE
[FLINK-25899] Add connected components example

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR adds an example for calculating the connected components using Stateful Functions.

This PR needs to be ported to `dev` and `release-3.2`.

### CODE REVIEW

**Review by igalshilman — APPROVED** [review]
Thanks @tillrohrmann! this is a very interesting example, and really shows the versatility of StateFun! 

I've left few comments for your consideration, mainly it is about the use of Java serialisation, and `Type`s in general.

File: java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/JavaType.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public Set<TypeCharacteristics> typeCharacteristics() {
+        return Collections.unmodifiableSet(EnumSet.of(TypeCharacteristics.IMMUTABLE_VALUES));
+    }
+
+    private static final class JavaSerializer<T> implements TypeSerializer<T> {
+        @Override
+        public Slice serialize(T value) {
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/JavaType.java:50]: Sounds good. Let's not give a wrong example. I'll remove the Java serialization.

File: java/connected-components/src/main/protobuf/messages.proto [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+syntax = "proto3";
+
+package org.apache.flink.statefun.playground.java.connectedcomponents.types;
+
+option java_package = "org.apache.flink.statefun.playground.java.connectedcomponents.types.generated";
+option java_multiple_files = true;
+
+message VertexComponentChange {
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply java/connected-components/src/main/protobuf/messages.proto:26]: I think this makes a lot of sense. I will remove the protobuf dependency.

File: java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/Types.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  private static final ObjectMapper JSON_OBJ_MAPPER = new ObjectMapper();
+  private static final String TYPES_NAMESPACE = "connected-components.types";
+
+  /**
+   * Type denoting a new vertex coming from the input source.
+   */
+  public static final Type<Vertex> VERTEX_INIT_TYPE =
+      SimpleType.simpleImmutableTypeFrom(
+          TypeName.typeNameOf(TYPES_NAMESPACE, Vertex.class.getName()),
```
[the remark this answers is not in the document]
Comment (tillrohrmann) [reply java/connected-components/src/main/java/org/apache/flink/statefun/playground/java/connectedcomponents/types/Types.java:24]: True. Yes, will use a string constant.

**Comment by tillrohrmann** [comment]
Thanks for the review @igalshilman. I've addressed your comments and pushed an update. Merging this PR now into `release-3.1`, `release-3.2` and `dev`.

**Pull request merged**